File: examples/Training/python/Read_Data.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
# Copyright (C) 2014 University of Dundee & Open Microscopy Environment.
#                    All Rights Reserved.
# Use is subject to license terms supplied in LICENSE.txt
#

"""
FOR TRAINING PURPOSES ONLY!
"""
from __future__ import division
from __future__ import print_function

from builtins import range
import omero
from omero.gateway import BlitzGateway
from Parse_OMERO_Properties import USERNAME, PASSWORD, HOST, PORT
from Parse_OMERO_Properties import datasetId, imageId, plateId

"""
start-code
"""

# Create a connection
# ===================
conn = BlitzGateway(USERNAME, PASSWORD, host=HOST, port=PORT)
conn.connect()
datasetId = int(datasetId)


def print_obj(obj, indent=0):
    """
    Helper method to display info about OMERO objects.
    Not all objects will have a "name" or owner field.
    """
    print("""%s%s:%s  Name:"%s" (owner=%s)""" % (
        " " * indent,
        obj.OMERO_CLASS,
        obj.getId(),
        obj.getName(),
        obj.getOwnerOmeName()))


# List all Projects owned by the user currently logged in
# =======================================================
# By default this returns Projects from all owners across
# all groups. We can filter by group and owner using the
# optional opts dict (new in 5.3.0)
# We also order by name and use 'limit' and 'offset',
# to load the first 5 Projects
print("\nList Projects:")
print("=" * 50)
my_exp_id = conn.getUser().getId()
default_group_id = conn.getEventContext().groupId
for project in conn.getObjects("Project", opts={'owner': my_exp_id,
                                                'group': default_group_id,
                                                'order_by': 'lower(obj.name)',
                                                'limit': 5, 'offset': 0}):
    print_obj(project)
    assert project.getDetails().getOwner().id == my_exp_id
    # We can get Datasets with listChildren, since we have the Project already.
    # Or conn.getObjects("Dataset", opts={'project', id}) if we have Project ID
    for dataset in project.listChildren():
        print_obj(dataset, 2)
        for image in dataset.listChildren():
            print_obj(image, 4)

# Find objects by ID. NB: getObjects() returns a generator, not a list
datasets = conn.getObjects("Dataset", [datasetId, datasetId+1])

# Get a single object by ID.
img = conn.getObject("Image", imageId)
img_name = img.getName()
# Can use "Annotation" for all types of annotations by ID
annotation = conn.getObject("Annotation", 1)

# Find an Object by attribute. E.g. 'name'
images = conn.getObjects("Image", attributes={"name": img_name})


# Get different types of Annotations
# ===========================

# Supported types are: ``tagannotation``, ``longannotation``,
# ``booleanannotation``, ``fileannotation``, ``doubleannotation``,
# ``termannotation``, ``timestampannotation``, ``mapannotation``

# List All Tags that you have permission to access
conn.getObjects("TagAnnotation")

# Find Tags with a known text value
tags = conn.getObjects("TagAnnotation", attributes={"textValue": "OK"})


# Retrieve 'orphaned' objects
# ===========================
# We can use the 'orphaned' filter to find Datasets, Images
# or Plates that are not in any parent container
print("\nList orphaned Datasets: \n", "=" * 50)
datasets = conn.getObjects("Dataset", opts={'orphaned': True})
for dataset in datasets:
    print_obj(dataset)


# Retrieve objects in a container
# ==========================================
# We can filter Images by their parent Dataset
# We can also filter Datasets by 'project', Plates by 'screen',
# Wells by 'plate'
print("\nImages in Dataset:", datasetId, "\n", "=" * 50)
for image in conn.getObjects('Image', opts={'dataset': datasetId}):
    print_obj(image)


# Retrieve an image by ID
# =======================
# Pixels and Channels will be loaded automatically as needed
image = conn.getObject("Image", imageId)
print("\nImage:%s" % imageId)
print("=" * 50)
print(image.getName(), image.getDescription())
# Retrieve information about an image.
print(" X:", image.getSizeX())
print(" Y:", image.getSizeY())
print(" Z:", image.getSizeZ())
print(" C:", image.getSizeC())
print(" T:", image.getSizeT())
# List Channels (loads the Rendering settings to get channel colors)
for channel in image.getChannels():
    print('Channel:', channel.getLabel(), end=' ')
    print('Color:', channel.getColor().getRGB())
    print('Lookup table:', channel.getLut())
    print('Is reverse intensity?', channel.isReverseIntensity())

# render the first timepoint, mid Z section
z = image.getSizeZ()// 2
t = 0
rendered_image = image.renderImage(z, t)
# renderedImage.show()               # popup (use for debug only)
# renderedImage.save("test.jpg")     # save in the current folder


# Get Pixel Sizes for the above Image
# ===================================
size_x = image.getPixelSizeX()       # e.g. 0.132
print(" Pixel Size X:", size_x)
if size_x:
    # Units support, new in OMERO 5.1.0
    size_x_obj = image.getPixelSizeX(units=True)
    print("Size X:", size_x_obj.getValue(), "(%s)" % size_x_obj.getSymbol())
    # To get the size with different units, e.g. Angstroms
    size_x_ang = image.getPixelSizeX(units="ANGSTROM")
    print("Size X:", size_x_ang.getValue(), "(%s)" % size_x_ang.getSymbol())


# Retrieve Screening data
# =======================
print("\nList Screens:")
print("=" * 50)
for screen in conn.getObjects("Screen"):
    print_obj(screen)
    for plate in screen.listChildren():
        print_obj(plate, 2)


# Retrieve Wells and Images within a Plate
# ========================================
if int(plateId) >= 0:
    print("\nPlate:%s" % plateId)
    print("=" * 50)
    plate = conn.getObject("Plate", plateId)
    print("\nNumber of fields:", plate.getNumberOfFields())
    print("\nGrid size:", plate.getGridSize())
    print("\nWells in Plate:", plate.getName())
    for well in plate.listChildren():
        index = well.countWellSample()
        print("  Well: ", well.row, well.column, " Fields:", index)
        for index in range(0, index):
            print("    Image: ",
                  well.getImage(index).getName(),
                  well.getImage(index).getId())


# List all annotations on an object. Filter for Tags and get textValue**

for ann in image.listAnnotations():
    print(ann.getId(), ann.OMERO_TYPE)
    print(" added by ", ann.link.getDetails().getOwner().getOmeName())
    if ann.OMERO_TYPE == omero.model.TagAnnotationI:
        print("Tag value:", ann.getTextValue())

# Get Links between Objects and Annotations
# Find Images linked to Annotation(s), unlink Images from these annotations
# and link them to another Tag Annotation
annotation_ids = [1, 2, 3]
tag_id = 4
for link in conn.getAnnotationLinks('Image', ann_ids=annotation_ids):
    print("Image ID:", link.getParent().id)
    print("Annotation ID:", link.getChild().id)
    # Update the child of the underlying omero.model.ImageAnnotationLinkI
    link._obj.child = omero.model.TagAnnotationI(tag_id, False)
    link.save()

# Find Annotations linked to Object(s), filter by namespace (optional)
for link in conn.getAnnotationLinks('Image', parent_ids=[1,2], ns="test.namespace"):
    print("Annotation ID:", link.getChild().id)


# Close connection
# ================
# When you are done, close the session to free up server resources.
conn.close()

File: examples/Training/matlab/ReadData.m
% Copyright (C) 2011-2014 University of Dundee & Open Microscopy Environment.
% All rights reserved.
%
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License along
% with this program; if not, write to the Free Software Foundation, Inc.,
% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

try
    % Initialize a client and a session using the ice.config file
    % See ConnectToOMERO for alternative ways to initialize a session
    [client, session] = loadOmero();
    p = parseOmeroProperties(client);
    eventContext = session.getAdminService().getEventContext();
    fprintf(1, 'Created connection to %s\n', p.hostname);
    msg = 'Created session for user %s (id: %g) using group %s (id: %g)\n';
    fprintf(1, msg, char(eventContext.userName), eventContext.userId,...
        char(eventContext.groupName), eventContext.groupId);
    
    % Information to edit
    projectId = p.projectid;
    datasetId = p.datasetid;
    imageId = p.imageid;
    plateId = p.plateid;

    print_object = @(x) fprintf(1, '  %s (id: %d, owner: %d, group: %d)\n',...
        char(x.getName().getValue()), x.getId().getValue(),...
        x.getDetails().getOwner().getId().getValue(),...
        x.getDetails().getGroup().getId().getValue());

    %% Projects
    % Retrieve all the projects and orphaned datasets owned by session
    % owner in the current context
    % If a project contains datasets, the datasets will automatically be
    % loaded but the images contained in the datasets are not loaded.
    disp(['Retrieving projects and orphaned datasets owned '...
        'by the session user in the current group']);
    [projects, orphanedDatasets] = getProjects(session);
    fprintf(1, '  Found %g projects\n', numel(projects));
    for i = 1 : numel(projects),
        print_object(projects(i));
        datasets = toMatlabList(projects(i).linkedDatasetList);
        for j = 1 : numel(datasets),
            fprintf(1, '  ');
            print_object(datasets(j));
        end
    end
    fprintf(1, '  Found %g orphaned datasets\n', numel(orphanedDatasets));
    for j = 1 : numel(orphanedDatasets),
        print_object(orphanedDatasets(j));
    end
    fprintf(1, '\n');

    % Retrieve all the unloaded projects owned by the session owner across
    % groups
    disp('Retrieving projects owned by the session user across groups')
    projects = getProjects(session, 'group', -1);
    fprintf(1, '  Found %g projects\n', numel(projects));
    for i = 1 : numel(projects),
        print_object(projects(i));
    end
    fprintf(1, '\n');

    % Retrieve all the unloaded projects owned by the session owner across
    % groups
    disp('Retrieving projects across by any user in the current group')
    projects = getProjects(session, 'owner', -1);
    fprintf(1, '  Found %g projects\n', numel(projects));
    for i = 1 : numel(projects),
        print_object(projects(i));
    end
    fprintf(1, '\n');

    % Retrieve a loaded project specified by an input identifier
    % If the dataset contains images, the images will be loaded
    fprintf(1, 'Reading project %g with loaded images\n', projectId);
    project = getProjects(session, projectId, true);
    assert(~isempty(project), 'OMERO:ReadData', 'Project Id not valid');
    datasets = toMatlabList(project.linkedDatasetList);
    for i = 1 : numel(datasets),
        print_object(datasets(i));
        images = toMatlabList(datasets(i).linkedImageList);
        for j = 1 : numel(images)
            fprintf(1, '  ');
            print_object(images(j));
        end
    end
    fprintf(1, '\n');

    %% Datasets
    % Retrieve all the unloaded datasets owned by the session owner.
    % If the datasets contain images, the images will not be loaded.
    disp('Listing datasets owned by the session user');
    allDatasets = getDatasets(session);
    fprintf(1, '  Found %g datasets\n', numel(allDatasets));
    for i = 1 : numel(allDatasets),
        print_object(allDatasets(i));
    end
    fprintf(1, '\n');

    % Retrieve all the unloaded datasets owned by the session owner across
    % all groups
    disp('Retrieving dataset owned by the session user across all groups')
    allDatasetsAllGroups = getDatasets(session, 'group', -1);
    fprintf(1, '  Found %g datasets\n', numel(allDatasetsAllGroups));
    for i = 1 : numel(allDatasetsAllGroups),
        print_object(allDatasetsAllGroups(i));
    end
    fprintf(1, '\n');

    % Retrieve a loaded dataset specified by an input identifier
    % If the dataset contains images, the images will be loaded
    fprintf(1, 'Retrieving dataset %g with loaded images\n', datasetId);
    dataset = getDatasets(session, datasetId, true);
    assert(~isempty(dataset), 'OMERO:ReadData', 'Dataset Id not valid');
    images = toMatlabList(dataset.linkedImageList); % The images in the dataset.
    fprintf(1, '  Found %g images\n', numel(images));
    for i = 1 : numel(images),
        print_object(images(i));
    end
    fprintf(1, '\n');

    %% Images
    % Retrieve all the images owned by the session user.
    disp('Retrieving images owned by the session user in the current group')
    allImages = getImages(session);
    fprintf(1, ' Found %g images\n', numel(allImages));
    for i = 1 : numel(allImages),
        print_object(allImages(i));
    end
    fprintf(1, '\n');

    % Retrieve all the images owned by the session user across all groups
    disp('Retrieving images owned by the session user across groups')
    allImagesAllGroups = getImages(session, 'group', -1);
    fprintf(1, '  Found %g images\n', numel(allImagesAllGroups));
    for i = 1 : numel(allImagesAllGroups),
        print_object(allImagesAllGroups(i));
    end
    fprintf(1, '\n');

    % Retrieve all the images contained in a given dataset.
    fprintf(1, 'Retrieving images contained in dataset %g\n', datasetId);
    images2 = getImages(session, 'dataset', datasetId);
    fprintf(1, '  Found %g images\n', numel(images2));
    for i = 1 : numel(images2),
        print_object(images2(i));
    end
    fprintf(1, '\n');

    % Retrieve all the images contained in a given dataset.
    fprintf(1, 'Retrieving images contained in project %g\n', projectId);
    images3 = getImages(session, 'project', projectId);
    fprintf(1, '  Found %g images\n', numel(images3));
    for i = 1 : numel(images3),
        print_object(images3(i));
    end
    fprintf(1, '\n');

    % Retrieve an image if the identifier is known.
    fprintf(1, 'Retrieving image: %g\n', imageId);
    image = getImages(session, imageId);
    proxy = session.getContainerService();
    assert(~isempty(image), 'OMERO:ReadData', 'Image Id not valid');
    
    % Access information about the image for example to draw it.
    
    % The model is a follow: Image-Pixels i.e. to access valuable data about
    % the image you need to use the pixels object.
    % We now only support one set of pixels per image (it used to be more!).
    
    pixels = image.getPrimaryPixels();
    sizeZ = pixels.getSizeZ().getValue(); % The number of z-sections.
    sizeT = pixels.getSizeT().getValue(); % The number of timepoints.
    sizeC = pixels.getSizeC().getValue(); % The number of channels.
    sizeX = pixels.getSizeX().getValue(); % The number of pixels along the X-axis.
    sizeY = pixels.getSizeY().getValue(); % The number of pixels along the Y-axis.
    fprintf(1, '  SizeX: %g\n', sizeX);
    fprintf(1, '  SizeY: %g\n', sizeY);
    fprintf(1, '  SizeZ: %g\n', sizeZ);
    fprintf(1, '  SizeC: %g\n', sizeC);
    fprintf(1, '  SizeT: %g\n', sizeT);
    fprintf(1, '\n');
    
    %% Screens
    % Retrieve Screening data owned by the user currently logged in.
    
    % There is no explicit method in the gateway exposed to retrieve screening data
    % (to learn about the model go to ScreenPlateWell) but you can use the ContainerService to
    % load the data, you can use the method `findAllByQuery`.

    % load Screen and plate owned by the user currently logged in
    disp('Retrieving all screens and orphaned plates owned by the session user')
    [screens, orphanedPlates] = getScreens(session);
    fprintf(1, '  Found %g screens\n', numel(screens));
    for i = 1 : numel(screens),
        print_object(screens(i));
        plates = toMatlabList(screens(i).linkedPlateList);
        for j = 1 : numel(plates),
            fprintf(1, '  ');
            print_object(plates(j));
            plateAcquisitions = toMatlabList(plates(j).copyPlateAcquisitions());
            for k = 1 : numel(plateAcquisitions),
                fprintf(1, '    ');
                print_object(plateAcquisitions(k));
            end
        end
    end
    fprintf(1, '  Found %g orphaned plates\n', numel(orphanedPlates));
    for i = 1 : numel(orphanedPlates),
        print_object(orphanedPlates(i));
        plateAcquisitions = toMatlabList(orphanedPlates(i).copyPlateAcquisitions());
        for j = 1 : numel(plateAcquisitions),
            fprintf(1, '  ');
            print_object(plateAcquisitions(j));
        end
    end
    fprintf(1, '\n');

    % Retrieve all the images owned by the session user across all groups
    disp('Retrieving all screens owned by the session user across all groups')
    allScreensAllGroups = getScreens(session, 'group', -1);
    fprintf(1, '  Found %g screens\n', numel(allScreensAllGroups));
    for i = 1 : numel(allScreensAllGroups),
        print_object(allScreensAllGroups(i));
    end
    fprintf(1, '\n');

    %% Plates
    % Retrieve all the plates owned by the session user.
    disp('Listing plates owned by the session user');
    allPlates = getPlates(session);
    fprintf(1, '  Found %g plates\n', numel(allPlates));
    for i = 1 : numel(allPlates),
        print_object(allPlates(i));
    end
    fprintf(1, '\n');

    % Retrieve all the pldates owned by the session user across all groups
    disp('Retrieving plates owned by the session user across all groups')
    allPlatesAllGroups = getPlates(session, 'group', -1);
    fprintf(1, '  Found %g plates\n', numel(allPlatesAllGroups));
    for i = 1 : numel(allPlatesAllGroups),
        print_object(allPlatesAllGroups(i));
    end
    fprintf(1, '\n');

    % Retrieve Wells within a Plate, see ScreenPlateWell.

    % Given a plate ID, load the wells.
    % You will have to use the findAllByQuery method.
    fprintf(1, 'Listing wells for plate: %g\n ', plateId);
    wells = session.getQueryService().findAllByQuery(['select well from Well as well left outer join fetch well.plate as pt left outer join fetch well.wellSamples as ws left outer join fetch ws.plateAcquisition as pa left outer join fetch ws.image as img left outer join fetch img.pixels as pix left outer join fetch pix.pixelsType as pt where well.plate.id =  ', num2str(plateId)], []);
    wells = toMatlabList(wells);
    fprintf(1, '  Found %g wells\n ', numel(wells));
    
    for i = 1:numel(wells),
        wellSamples = toMatlabList(wells(i).copyWellSamples());
        fprintf(1, '  Well %g - Found %g well samples\n ', i, numel(wellSamples));
        for j = 1:numel(wellSamples),
            pa = wellSamples(j).getPlateAcquisition();
        end
    end
    
catch err
    client.closeSession();
    throw(err);
end

% Close the session
client.closeSession();
